libraries/ms-helper/src/persist.c

```diff
@@ -0,0 +1,158 @@
+  }
+
+  persist->blob = blob;
+  persist->blob_size = blob_size;
+  persist->prev_flash_addr = PERSIST_INVALID_ADDR;
+
+  // Load stored data
+  PersistHeader header = {
+    .marker = PERSIST_VALID_MARKER,  //
+    .size_bytes = 0                  //
+  };
+  persist->flash_addr = PERSIST_BASE_ADDR;
+
+  // Essentially, we want to exit in three scenarios:
+  // * The page has been erased and there are no valid sections. Use the base address.
+  //   Marker == 0xFFFFFFFF, Size == 0xFFFFFFFF
+  // * This is an invalid section. Increment the address by (header + blob).
+  //   Marker != 0xFFFFFFFF, Size != 0xFFFFFFFF
+  // * This is a valid section. Use this address.
+  //   Marker == 0xFFFFFFFF, Size != 0xFFFFFFFF
+  do {
+    status_ok_or_return(
+        flash_read(persist->flash_addr, sizeof(header), (uint8_t *)&header, sizeof(header)));
+    if (header.marker != PERSIST_VALID_MARKER) {
+      persist->flash_addr += sizeof(header) + header.size_bytes;
+    }
+
+    if (persist->flash_addr >= PERSIST_END_ADDR) {
+      // Somehow had zero valid sections remaining - erase page
+      LOG_DEBUG("Somehow started with a full invalid page - erasing page\n");
+      status_ok_or_return(flash_erase(PERSIST_FLASH_PAGE));
+      persist->flash_addr = PERSIST_BASE_ADDR;
+    }
+  } while (header.marker != PERSIST_VALID_MARKER);
+
+  if (header.size_bytes == PERSIST_INVALID_SIZE) {
+    LOG_DEBUG("No valid sections found! New persist data will live at 0x%" PRIx32 "\n",
+              (uint32_t)persist->flash_addr);
+    persist_commit(persist);
+  } else if (header.size_bytes != persist->blob_size) {
+    LOG_DEBUG("Mismatched blob sizes! Invalidating old section\n");
+    persist->prev_flash_addr = persist->flash_addr;
+    persist->flash_addr += sizeof(header) + header.size_bytes;
+    persist_commit(persist);
+  } else {
+    LOG_DEBUG("Found valid section at 0x%" PRIx32 " (0x%" PRIx32 " bytes), loading data\n",
+              (uint32_t)persist->flash_addr, (uint32_t)header.size_bytes);
+    StatusCode ret = flash_read(persist->flash_addr + sizeof(header), persist->blob_size,
+                                (uint8_t *)persist->blob, persist->blob_size);
+    status_ok_or_return(ret);
+
+    // Increment flash_addr to the next new section
+    persist->prev_flash_addr = persist->flash_addr;
+    persist->flash_addr += sizeof(header) + header.size_bytes;
+  }
+
+  return soft_timer_start_millis(PERSIST_COMMIT_TIMEOUT_MS, prv_periodic_commit, persist, NULL);
+}
+
+StatusCode persist_commit(PersistStorage *persist) {
+  // Mark previous section as invalid
+  if (persist->prev_flash_addr != PERSIST_INVALID_ADDR) {
+    uint32_t invalid = 0;
+    flash_write(persist->prev_flash_addr, (uint8_t *)&invalid, sizeof(invalid));
+  }
+
+  // Check if we're overrunning the page
+  if (persist->flash_addr + sizeof(PersistHeader) + persist->blob_size > PERSIST_END_ADDR) {
+    // It should be okay if we're right on the boundary?
```

Comment:
Maybe write a test for this if you aren't sure :p

Reply:
:P

Reply:
I would still say test it at least once manually so we are certain :p